rfd/0050-join-methods.md

```diff
@@ -0,0 +1,109 @@
+TLS certificates to the node for the requested role (Proxy, Node, Kubernetes,
+etc).
+
+The joining node also needs to be able to trust that the Teleport cluster is
+authentic. It can do this by using CA pins configured on the node, in the case
+of IoT nodes connecting to the proxy, the proxy TLS cert can be trusted through
+PKI.
+
+### Current Join Endpoints
+
+#### Auth `POST /tokens/register`
+Join methods: token, EC2
+Server trust: CA pins
+This is the main token register endpoint on the Auth server. Accepts tokens or
+EC2 Identity Documents to authenticate the client.
+
+#### Auth `POST /tokens/register/auth`
+Join Methods: token
+Server trust: CA pins
+Looks like this is a legacy endpoint the just deletes the given token. Only
+works for tokens with the `Auth` role. I can't find anything that calls this, so
+it can probably be scheduled for deletion.
+
+#### Auth `rpc GenerateHostCerts(HostCertsRequest) returns (Certs)`
+Client trust: mTLS
+Server trust: mTLS
+This is an mTLS authenticated gRPC endpoint for cert renewal. The node must
+first join the cluster using another endpoint to get its first client
+certificate used for the mTLS connection.
+
+#### Proxy `POST /webapi/host/credentials`
+Join methods: token, EC2
+Server trust: PKI
+This is the proxy endpoint for registering IoT nodes, it basically forwards to
+the Auth `/tokens/register` endpoint.
+
+### New IAM Join Method
+The IAM join method requires new gRPC methods because the design requires gRPC
+streams to implement a challenge/response protocol. A new gRPC method will be
+added on Auth and Proxy to support this. The client will be able to call the
+gRPC method with either a Proxy or Auth address and it will "just work"
+transparently.
+
+#### Auth `rpc RegisterUsingIAM(stream RegisterUsingIAMRequest) returns (stream RegisterUsingIAMResponse)`
+Join mathods: IAM
+Server trust: CA pins
+This is the Auth endpoint that will complete the IAM join request.
+
+#### Proxy `rpc RegisterUsingIAM(stream RegisterUsingIAMRequest) returns (stream RegisterUsingIAMResponse)`
+Join mathods: IAM
+Server trust: PKI
+Normally, authenticated gRPC calls from IoT nodes are tunnelled through the
+proxy over SSH. This will not work for unauthenticated clients which don't yet
+have an SSH certificate.
+
+It would be possible to use TLS routing on the proxy to forward gRPC requests
+directly to the Auth server, but this would open up the Auth server to DOS
+attacks from unauthenticated clients.
+
+Instead, this single gRPC method will be implemented on the proxy. This method
+will first perform rate-limiting, then "forward" the request to the auth server
+method at the application layer by calling the gRPC method through the Proxy's
+own Auth client.
+
+Since the Proxy currently does not expose any gRPC service of its own, a new
+gRPC listener will be added to support this. gRPC connections will be
+multiplexed on the web listener/port using the existing ALPN-based multiplexer
```

Comment:
Love this idea.

Does the ALPN multiplexer have to be explicitly enabled? If so I would be more explicit and list that as a prerequisite in order to support IAM joining.

Reply:
No, the ALPN multiplexer is enabled by default, independent of the configured ProxyListenerMode (that's just a matter of whether the proxy listens on one or multiple ports, the multiplexer is enabled either way).

I think it makes sense to force this to use the web port rather than giving the option of putting it on another port, and have it always on independent of the ProxyListenerMode